fl4health/clients/basic_client.py

```diff
@@ -1206,3 +1276,55 @@ def transform_gradients(self, losses: TrainingLosses) -> None:
             losses (TrainingLosses): The losses object from the train step
         """
         pass
+
+    def save_checkpoint(self) -> None:
+        """
+        Saves checkpoint dict consisting of client name, total steps, lr schedulers,
+            metrics reporter and optimizers state. Method can be overridden to augment saved checkpointed state.
+        """
+
+        assert self.per_round_checkpointer is not None
+
+        ckpt = {
+            "lr_schedulers_state": {key: scheduler.state_dict() for key, scheduler in self.lr_schedulers.items()},
+            "total_steps": self.total_steps,
+            "client_name": self.client_name,
+            "metrics_reporter": self.metrics_reporter,
+            "optimizers_state": {key: opt.state_dict()["state"] for key, opt in self.optimizers.items()},
+        }
+
+        self.per_round_checkpointer.save_checkpoint(ckpt)
+
+        log(INFO, f"Saving client state to checkpoint at {self.per_round_checkpointer.checkpoint_path}")
+
+    def load_checkpoint(self) -> None:
+        """
+        Load checkpoint dict consisting of client name, total steps, lr schedulers, metrics
+            reporter and optimizers state. Method can be overriden to augment loaded checkpointed state.
+        """
+        assert self.per_round_checkpointer is not None and self.per_round_checkpointer.checkpoint_exists()
+
+        ckpt = self.per_round_checkpointer.load_checkpoint()
+
+        assert "lr_schedulers_state" in ckpt and isinstance(ckpt["lr_schedulers_state"], dict)
+        assert "client_name" in ckpt and isinstance(ckpt["client_name"], str)
+        assert "total_steps" in ckpt and isinstance(ckpt["total_steps"], int)
+        assert "metrics_reporter" in ckpt and isinstance(ckpt["metrics_reporter"], MetricsReporter)
+        assert "optimizers_state" in ckpt and isinstance(ckpt["optimizers_state"], dict)
```

Comment:
The `narrow_config_type` function, essentially does this type of thing. Maybe we should name it something more general and use it here, since it's just effectively working on a dictionary at the end of the day?

Reply:
Good idea!